File: ruby/stack.rb
class Stack

  attr_reader :items, :size

  def initialize(items = {}, size = 0)
    @items, @size = items, size
  end


  def push item
    if @size == 0
      @items[0] = item
      @size = @size + 1
    else
      all_items = @items.values
      all_items.unshift(item)
      @size = @size + 1

      all_items.each_with_index do |el, i|
        @items[i] = el
      end
    end
  end


  def pop
    top_object = ''
    all_items = @items.values
    top_object = all_items[0]
    ze_items = all_items[1..all_items.length]

    ze_items.each_with_index do |el, i|
      @items[i] = el
    end


    @items.delete(@size - 1)
    @size = @size - 1
    top_object
  end


  def peek
    @items[0]
  end


  def is_empty?
    @size  === 0 ? true : false
  end


  def clear
    @items = {}
    @size = 0
  end


  def stack_size
    @size
  end


  def to_array
    @items.values
  end

end

File: javascript/stack.js
//Last In Last Out

let Stack = function() {
  this.items = {};
  this.size = 0;
}


//Push Function
//Inserts a new object
//to the top of the stack
//without returning it

Stack.prototype.push = function(item){
  if(this.size === 0) {
    this.items[this.size] = item;
    this.size++;

  } else {
    let allItems = Object.keys(this.items).map(el => this.items[el]);
    allItems.unshift(item);
    this.size++;

    for(let i = 0; i < allItems.length; i++){
      this.items[i] = allItems[i]
    }
  }
}


//Pop Function
//Removes and returns the
//object at the top of the stack

Stack.prototype.pop = function(){
  let topObject;

  let allItems = Object.keys(this.items).map((el, i) => {
    if(i !== 0){
      return this.items[el]
    } else {
      topObject = this.items[el];
    }
  });

  allItems.shift();

  for(let z = 0; z < allItems.length; z++){
    this.items[z] = allItems[z];
  }

  delete this.items[this.size - 1];
  this.size--;

  return topObject;
}


//Peek Function
//Returns the object at the top
//of the stack without removing
//it

Stack.prototype.peek = function(){
  return this.items[0];
}


//ToArray Function
//Copies the stack to a new Array

Stack.prototype.ToArray = function(){
  return Object.keys(this.items).map(el => this.items[el]);
}


//isEmpty Function
//Checks to see if stack is
//empty

Stack.prototype.isEmpty = function(){
  return this.size === 0 ? true : false;
}


//Clear Function
//Removes all objects from
//The stack
Stack.prototype.clear = function(){
  this.items = {};
  this.size = 0;
}


//StackSize Function
//Returns the size of the
//current stack

Stack.prototype.stackSize = function(){
  return this.size;
}
